### PR TITLE
sway: 1.5 -> 1.6; wlroots: 1.12 -> 1.13

### PR DESCRIPTION
**Pull request by Synthetica9**
###### Motivation for this change

Tracking PR for Sway 1.6, first RC just landed: https://github.com/swaywm/sway/releases/tag/1.6-rc1

EDIT 2021-03-24: Next RC landed https://github.com/swaywm/sway/releases/tag/1.6-rc2

EDIT: 2021-04-07: Final version of wlroots 0.13 landed. https://github.com/swaywm/wlroots/releases/tag/0.13.0

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Synthetica9** [comment]
Seems to build and run fine here. A few new dependencies were added, I just naively added them.

File: pkgs/applications/window-managers/sway/default.nix
Comment (primeos) [inline pkgs/applications/window-managers/sway/default.nix:41]: We should avoid that. Features that we don't want should be disabled explicitly.

File: pkgs/development/libraries/wlroots/default.nix
Comment (primeos) [inline pkgs/development/libraries/wlroots/default.nix:14]: Nitpick (this'll be squashed later) but I'd prefer to use the full SHA-1 hashes / commit IDs here.

**Review by primeos — CHANGES_REQUESTED** [review]
Can you also run `nixpkgs-review` and investigate broken builds (e.g. by backporting patches, switching to a newer release, looking at issue trackers, etc.)?
Other than that it looks good.

File: pkgs/applications/window-managers/sway/default.nix
Comment (primeos) [reply pkgs/applications/window-managers/sway/default.nix:41]:
```suggestion
    "-Dsd-bus-provider=libsystemd"
```

**Comment by jbeich** [comment]
For wayfire/hikari/etc fixes try extracting from https://github.com/freebsd/freebsd-ports/commit/79579d5405fe38a2f8c8956e6f3b3f4545e058db

File: pkgs/applications/window-managers/sway/default.nix
Comment (AndersonTorres) [inline pkgs/applications/window-managers/sway/default.nix:22]: Is this a Nixpkgs-only patch?

File: pkgs/development/libraries/wlroots/default.nix
Comment (AndersonTorres) [inline pkgs/development/libraries/wlroots/default.nix:30]: Use `substituteInPlace` here.

(To be honest I also prefer sed, but using the nixpkgs-provided functions is the preferrable way.)

**Review by AndersonTorres — DISMISSED** [review]

File: pkgs/development/libraries/wlroots/default.nix
Comment (AndersonTorres) [inline pkgs/development/libraries/wlroots/default.nix:26]: put `xcbutilrenderutil` explicitly as dependency (and use `inherit (xorg) xcbutilrenderutil;` in the callPackage).

File: pkgs/development/libraries/wlroots/default.nix
Comment (Synthetica9) [reply pkgs/development/libraries/wlroots/default.nix:30]: This is going to be squashed away when wlroots 0.13 is released

File: pkgs/applications/window-managers/sway/default.nix
Comment (Synthetica9) [reply pkgs/applications/window-managers/sway/default.nix:22]: Yes, but it will also be squashed away

File: pkgs/applications/window-managers/labwc/default.nix
Comment (Synthetica9) [reply pkgs/applications/window-managers/labwc/default.nix:48]: Only scdoc I think?

File: pkgs/applications/window-managers/dwl/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/dwl/default.nix:23]: Please don't fetch PRs but commits instead.

File: pkgs/applications/window-managers/sway/default.nix
Comment (SuperSandro2000) [reply pkgs/applications/window-managers/sway/default.nix:22]: Such version ranges changes should use substituteInPlace.

File: pkgs/applications/window-managers/labwc/default.nix
Comment (AndersonTorres) [reply pkgs/applications/window-managers/labwc/default.nix]: Yes, only scdoc

File: pkgs/applications/window-managers/sway/default.nix
Comment (AndersonTorres) [reply pkgs/applications/window-managers/sway/default.nix:22]: Do this via scripting, it will be easier to undo later.

**Comment by Synthetica9** [comment]
Sway 1.6 also seems to fix an issue I was having with firefox, which is a nice plus

File: pkgs/development/libraries/wlroots/default.nix
Comment (primeos) [reply pkgs/development/libraries/wlroots/default.nix]: @AndersonTorres do we have a convention for that? I was a bit surprised about this but then again it shouldn't really matter anyway.

File: pkgs/applications/window-managers/waybox/default.nix
Comment (primeos) [inline pkgs/applications/window-managers/waybox/default.nix:44]: We should also notify the maintainers if we mark something as broken.
cc @AndersonTorres (already in this thread but to make it explicit)

I'd assume that `waybox` is still in use so we should either override `wlroots` (0.12) or add a TODO comment if this is only temporary and will be dropped before this PR gets merged.

Comment (Synthetica9) [reply pkgs/applications/window-managers/waybox/default.nix:44]: There doesn't seem to be any upstream development in the past half year or so, so I marked it as broken for now. I'll add a note, and I guess we'll see what the status is when the final version of wlroots 0.13 is released.

Comment (primeos) [reply pkgs/applications/window-managers/waybox/default.nix:46]: Ok, that makes sense then :) Thanks!

**Comment by Synthetica9** [comment]
@SuperSandro2000 What's the recommended way to fetch multiple commits from a PR that hasn't been merged yet? (https://github.com/Hjdskes/cage/pull/174)

**Comment by Synthetica9** [comment]
@berbiche see 732cd86

**Comment by primeos** [comment]
> @SuperSandro2000 What's the recommended way to fetch multiple commits from a PR that hasn't been merged yet? (Hjdskes/cage#174)

I think that's unfortunately a bit of a problem without a good solution (PRs are mutable so we cannot use them directly and even fetching the commits before they're merged is far from ideal as they could disappear / become GCed, e.g. if the PR branch will be rebased).

For past wlroots updates I've kindly pinged the maintainers to e.g. merge such PRs. If that doesn't work it's probably best to keep wlroots 1.12 in Nixpkgs for some time (we're already doing this with some other packages).

**Comment by Synthetica9** [comment]
Okay, I see you pinged the maintainers. Since this is not yet ready to be merged anyways I'll keep it as is for now, I'll use the merge commit when that appears on master.

**Comment by SuperSandro2000** [comment]
GCed commits is usually not a problem with github and is far less likely than a changed or.

**Comment by berbiche** [comment]
> @berbiche see [732cd86](https://github.com/NixOS/nixpkgs/commit/732cd861f015d777756507a853d9dce806f74eda)

Cagebreak's source code history is available on the "development" branch, the master branch is just a "dump" like you said.

I'll investigate and send a patch when I have time.

File: pkgs/applications/window-managers/labwc/default.nix
Comment (jbeich) [inline pkgs/applications/window-managers/labwc/default.nix:34]: `pandoc`  isn't used after https://github.com/johanmalm/labwc/commit/5a36fef547fc

File: pkgs/applications/window-managers/labwc/default.nix
Comment (jbeich) [inline pkgs/applications/window-managers/labwc/default.nix:30]: Don't you need to bump `version` from `unstable-2021-02-06` to `unstable-2021-03-15` as well?
Alternatively, consider using `0.1.0` tag then applying https://github.com/johanmalm/labwc/pull/26.

**Comment by Synthetica9** [comment]
@alyssais I tried to override wlroots for wayfire and wcm, but I couldn't get it through the plugin system, any pointers?

**Comment by Synthetica9** [comment]
@primeos Should I do an initial squash, or wait with that until we're ready to merge?

**Comment by primeos** [comment]
@Synthetica9 feel free to already squash/rebase it if you have time. That way we might be able to merge it a bit faster and the PR would look a bit "cleaner" but I'm fine with it either way.

**Comment by Synthetica9** [comment]
> feel free to already squash\rebase it if you have time.

Okay, I'll see what I can do

> **Edit: Oh, bad bot... :o :D**

🤭

File: pkgs/applications/window-managers/sway/load-configuration-from-etc.patch
Comment (primeos) [inline pkgs/applications/window-managers/sway/load-configuration-from-etc.patch:43]:
```suggestion
+		{ .prefix = NIX_SYSCONFDIR, .config_folder = "sway"},
```

Didn't test this (yet) but this diff suggests it and https://github.com/swaywm/sway/commit/8f5c546fcadbacae5f9828bc1bfac247186cfdfc confirms it (the new `config_path` function).

**Comment by alyssais** [comment]
> @alyssais I tried to override wlroots for wayfire and wcm, but I
> couldn't get it through the plugin system, any pointers?

I've just pushed a commit to this PR that should do this -- do with it
whatever you want. :)

BTW: in my experience usual convention is to call versioned attributes
e.g. wlroots_0_12, rather than wlroots-0-12.

**Comment by alyssais** [comment]
I pushed a small cleanup to the Wayfire override.
This is all going to be squashed a bit before being merged, right?

**Comment by Synthetica9** [comment]
> This is all going to be squashed a bit before being merged, right?

Yeah, I'll do a rebase to group relevant commits together at least, maybe a full squash if requested.

**Comment by mweinelt** [comment]
https://github.com/swaywm/sway/releases/tag/1.6

**Comment by Synthetica9** [comment]
> [swaywm/sway@`1.6` (release)](https://github.com/swaywm/sway/releases/tag/1.6)

Way ahead of you 😉

**Comment by Synthetica9** [comment]
@primeos I think two things remain: 

- [ ] What do we do with https://github.com/Hjdskes/cage/pull/174 ? Pin wlroots to 0.12 for cage? 
- [ ] Rebase again, should I squash all RCs together or keep them for the "story"?

**Comment by primeos** [comment]
@Synthetica9 yes, I'd pin cage to wlroots 0.12 (we can always look at that later, outside of this PR) and squash the RCs together. Regarding squashing I'd do the following (IIRC that's how I handled it in the past but that is open for discussion if you don't like it):
1. Commit: Introduces `wlroots_0_12` and pins the packages that still require it to that version
2. Commit: Updates `wlroots` and introduces the required changes to fix builds of other packages (except if we update their version)
3. Commit: Updates `sway`

**Comment by Synthetica9** [comment]
I think this PR is ready to be merged :)

File: pkgs/applications/window-managers/cage/default.nix
Comment (primeos) [inline pkgs/applications/window-managers/cage/default.nix:7]: Nitpick but could you drop this leftover if you don't mind?

File: pkgs/top-level/all-packages.nix
Comment (primeos) [inline pkgs/top-level/all-packages.nix:26448]: We should re-add `recurseIntoAttrs` for `wayfireApplications-unwrapped` and `wayfirePlugins` (relevant for nix-env, the package search, etc.). That should also fix this nixpkgs-review output:
```
3 packages removed:
wayfire (†0.7.0) wcm (†0.7.0) wf-shell (†0.7.0)
```

Comment (primeos) [reply pkgs/top-level/all-packages.nix:26448]: If you don't mind you could also drop `being cage and wayfire.` from this commit message (there are more affected packages now but we don't need to name them in the commit message).

(The Wayfire change could be an extra commit as we're also making it extensible but that's likely not worth the effort so feel free to ignore that, depending on what you prefer.)

**Review by primeos — CHANGES_REQUESTED** [review]
Overall it LGTM now, thanks a lot! We mainly need to re-add `recurseIntoAttrs` for Wayfire (s. below).

`nixpkgs-review` result:
```
1 package added:
wlroots_0_12 (init at 0.12.0)

11 packages updated:
dwl fuzzel sway-contrib.grimshot (1.5.1 → 1.6) labwc (2021-02-06 → 2021-03-15) qt-video-wlr sway (1.5.1 → 1.6) sway-contrib.inactive-windows-transparency (1.5.1 → 1.6) sway-unwrapped (1.5.1 → 1.6) waybar wbg-unstable wlroots (0.12.0 → 0.13.0)

3 packages removed:
wayfire (†0.7.0) wcm (†0.7.0) wf-shell (†0.7.0)

12 packages built:
dwl fuzzel labwc qt-video-wlr sway sway-contrib.grimshot sway-contrib.inactive-windows-transparency sway-unwrapped waybar wbg wlroots wlroots_0_12
```

File: pkgs/top-level/all-packages.nix
Comment (Synthetica9) [reply pkgs/top-level/all-packages.nix]: @primeos I have pushed something that compiles, checking the result of nixpkgs-review now

Comment (Synthetica9) [reply pkgs/top-level/all-packages.nix]: Hmm, now it still says 

```
1 package removed:
wf-shell (†0.7.0)
```

Not sure why...

Comment (primeos) [reply pkgs/top-level/all-packages.nix]: It still missing for `wayfirePlugins` (will appear as `wayfireApplications-unwrapped.wayfirePlugins.wf-shell` in the search). Then it should work.

Comment (Synthetica9) [reply pkgs/top-level/all-packages.nix]: Oh, I see... Should it be on both or only on unwrapped?

Comment (primeos) [reply pkgs/top-level/all-packages.nix]: It used to be on both `wayfireApplications-unwrapped` and `wayfirePlugins` (s. diff) which is why I'd keep it on both.

Comment (Synthetica9) [reply pkgs/top-level/all-packages.nix]: Okay. Marking this resolved 👍🏻

**Review by primeos — APPROVED** [review]
Thanks a lot, very well done!
I've briefly tested `sway`, `dwl`, and `labwc` on my old desktop PC. `nixpkgs-review` result:
```
1 package added:
wayfireApplications-unwrapped.wlroots (init at 0.12.0)

11 packages updated:
dwl fuzzel sway-contrib.grimshot (1.5.1 → 1.6) labwc (2021-02-06 → 2021-03-15) qt-video-wlr sway (1.5.1 → 1.6) sway-contrib.inactive-windows-transparency (1.5.1 → 1.6) sway-unwrapped (1.5.1 → 1.6) waybar wbg-unstable wlroots (0.12.0 → 0.13.0)

12 packages built:
dwl fuzzel labwc qt-video-wlr sway sway-contrib.grimshot sway-contrib.inactive-windows-transparency sway-unwrapped waybar wayfireApplications-unwrapped.wlroots wbg wlroots
```

**Pull request merged**
